pkg/datapath/linux/devices.go

```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+// of devices is sent on the returned channel.
+func (dm *DeviceManager) Listen(ctx context.Context) (chan []string, error) {
+	return dm.listen(ctx, nil)
+}
+
+// listen is the internal method to start listening for changes to network devices. This is split from the
+// public method to allow tests to listen to devices in specific network namespace.
+func (dm *DeviceManager) listen(ctx context.Context, netNS *netns.NsHandle) (chan []string, error) {
+	l3DevOK := supportL3Dev()
+	routeChan := make(chan netlink.RouteUpdate)
+	closeChan := make(chan struct{})
+
+	err := netlink.RouteSubscribeWithOptions(routeChan, closeChan,
+		netlink.RouteSubscribeOptions{
+			Namespace: netNS,
+			// List existing routes to make sure we did not miss changes that happened after Detect().
+			// FIXME(JM): Perhaps could consider combining Detect() and Listen() so we don't have to do this.
+			// Having this whole thing be fully event-driven is likely a better design as that forces consumers
+			// to work with it correctly.
+			ListExisting: true,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	linkChan := make(chan netlink.LinkUpdate)
+	err = netlink.LinkSubscribeWithOptions(linkChan, closeChan, netlink.LinkSubscribeOptions{
+		Namespace: netNS,
+		// FIXME(JM): What if a device is removed between Detect() and Listen()?
+		ListExisting: false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	devicesChan := make(chan []string)
+
+	go func() {
+		// If a specific namespace is requested, lock the thread and set the
+		// threads namespace. This is currently only relevant for testing.
+		if netNS != nil {
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
+			if err := netns.Set(*netNS); err != nil {
+				log.WithError(err).Warnf("Failed to set network namespace")
+			}
+		}
+
+		// To avoid multiple reloads of the datapath when lots of routes are added in one go,
+		// collect the route updates into a batch and process the batch periodically.
+		//
+		ticker := time.NewTicker(dm.routeBatchingInterval)
+		defer ticker.Stop()
+		buffer := make([]netlink.Route, 0)
+
+		log.Printf("Listening for device changes")
+
+		for {
+			devicesChanged := false
+
+			select {
+			case <-ctx.Done():
+				log.Debug("context closed, Listen() stopping")
+				close(closeChan)
+				close(devicesChan)
+				return
+
+			case <-ticker.C:
+				if len(buffer) > 0 {
+					dm.Lock()
+					devicesChanged = dm.updateDevicesFromRoutes(l3DevOK, buffer)
+					dm.Unlock()
+					buffer = make([]netlink.Route, 0)
+				}
+
+			case update := <-routeChan:
+				if update.Type == unix.RTM_NEWROUTE {
+					buffer = append(buffer, update.Route)
+					ticker.Reset(dm.routeBatchingInterval)
+				}
+
+			case update := <-linkChan:
+				dm.Lock()
+				devicesChanged = dm.handleLinkUpdate(update)
+				dm.Unlock()
+			}
+
+			if devicesChanged {
+				devices := dm.GetDevices()
+				log.WithField(logfields.Devices, devices).Info("Detected new devices")
```

Comment:
We will already log that we detected the individual device (previous comment). Do we need both?

Also, this can be triggered from link deletion, so the log message is not exactly accurate here.

Reply:
Removed the comment for the individual device and left this one here. Changed it to "Detected devices".